### ISSUE TITLE
Linear congruential generator produces different values than gcc and msvc

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#33554]
|  |  |
| --- | --- |
| Bugzilla Link | [34206](https://llvm.org/bz34206) |
| Version | 5.0 |
| OS | MacOS X |
| Attachments | [linear_congruential_engine test](https://user-images.githubusercontent.com/60944935/143756159-96d3ccdd-489d-47dd-83b5-3fdd75d8b6aa.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @mclow |

## Extended Description 

To my understanding the C++ standard does not mandate the algorithms for random number distributions but it does mandate the algorithms for random number generators like linear_congruential_engine. So one should be able to generate consistent random values across all platforms/compilers.

However, i stumbled upon inconsistent values between various compilers and verified this observation with online compilers. I am not sure which compiler is right, but at least gcc and msvc agree on their results.

Please find attached a code snippet to reproduce the observed behavior.

I put this code snippet into Rextester:

  http://rextester.com/l/cpp_online_compiler_gcc (g++ 5.4.0)
  http://rextester.com/l/cpp_online_compiler_clang (clang 3.8.0)
  http://rextester.com/l/cpp_online_compiler_visual (Microsoft 19.00.23506 for x86)

Output:

  25214903928 206026503483683 245470556921330 (gcc)
  25214903928 18444698300399350051 8295313034219953650 (clang)
  25214903928 206026503483683 245470556921330 (msvc)

I also put this code snippet into a Compiler Explorer project. It illustrates that Clang 3.8 and Clang trunk produce exactly the same code/values.

  https://godbolt.org/g/rBSjaS

A possibly somehow related issue:

  llvm/llvm-project#28213

**Comment by jwakely** [comment]
The gzipped attachment is a small code snippet which might as well be pasted inline here:

```c++
#include <random>
#include <iostream>

int main()
{
  std::linear_congruential_engine<std::uint_fast64_t, 0x5DEECE66Dull, 0xBull, 1ull << 48> gen;

  for (int n=0; n<3; ++n)
    std::cout << gen() << ' ';
}
```

**Comment by LRFLEW** [comment]
If that code snippet is accurate, then it looks like this bug may have been resolved at some point. Godbolt has clang (trunk) producing the same output as the other implementations: https://godbolt.org/z/oY4odbYKT

I did make #81080, which should also fix incorrect LCG output, but it looks like it may be unrelated to this *particular* example of incorrect output.

**Comment by LRFLEW** [comment]
It looks like this specific instance of `std::linear_congruential_engine` returning the wrong values might have been fixed with LLVM 12. https://godbolt.org/z/vjr8MPWKW I've had issues checking libc++ with Godbolt in the past, so I'm not certain if this is 100% accurate, but I can't replicate the issue locally either.

**Comment by LRFLEW** [comment]
I found another example where libc++ produces the wrong values:

```
#include <random>
#include <iostream>

int main()
{
    std::linear_congruential_engine<std::uint64_t, 0x26b180695f9117ull, 0x12347ull, 3ull << 56> gen;

    for (int n=0; n<3; ++n)
        std::cout << gen() << ' ';
    std::cout << std::endl;

    return 0;
}
```
https://godbolt.org/z/5EaaTTfad

The third output value onwards differs in libc++ due to integer overflow. This example should be fixed in #81080.

**Comment by jwakely** [comment]
See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87744 for another case where libc++ gives incorrect results (I don't know if it's related to #81080)

```c++
#include <random>
#include <iostream>
int main() {
    std::linear_congruential_engine<unsigned long long int, 864691128455135232ULL, 12347ULL, 4052555153018976267ULL> gen;
    std::cout << gen() << '\n' << gen() << '\n' << gen() << '\n';
}
```

Libstdc++ and Boost.Random give:

864691128455147579
20120904253298372
499698276788149646

But clang 16 with libc++ gives:

864691128455147579
1965622972376959002
4035225266123976763

**Comment by jwakely** [comment]
Oh, I've just realised who @LRFLEW is :joy_cat:

**Comment by LRFLEW** [comment]
> Edit: I'm just repeating what the comment above already says, so removed.

I looked at the original comment, and it was different *enough* that it's still worth mentioning. For the record the comment brought up that *this* construction also gives incorrect values in libc++: `std::linear_congruential_engine<unsigned long long int, 864691128455135232ULL, 12347ULL, 4052555153018976267ULL>`.

This is slightly different from the example I gave above in terms of *why* libc++ is giving the wrong result. The example in my previous comment was caused by libc++ recognizing it *could* use Schrage's algorithm, but incorrectly thinking it didn't need to. **This** example is caused by libc++ recognizing that Schrage's algorithm cannot be used here, but due to a broken static assert, ends up using the standard `(ax+c)%m` implementation, which produces the wrong results due to integer overflow. Both of these are addressed in #81080, though the latter simply will refuse to compile. Getting that to compile should probably be its own PR, though.

And yes @jwakely, I figured it was you from the GCC bug report because of your username. I ended up here because the mention of #28213 caused me to realize that issue was supposedly fixed, and looking into how it was addressed led me to finding the problems I'm addressing with #81080.
